Add PDF documentation in the ISO

slaperche-scality opened the pull request:
**Component**:

build

**Context**: 

Add the PDF documentation in the ISO we ship

**Summary**:

- add tasks to build the documentation (`docs:html` and `docs:pdf`)
- add task to copy the generated PDF on the ISO

**Acceptance criteria**: 

When running `doit.sh -n4`, you will have `user-guide.pdf` under `_build/root/documentation`.

---

Closes: #1695

slaperche-scality committed:
build/doc: add tasks to build documentation

Add task to build documentation as HTML and PDF.

Refs: #1695
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/doc: add documentation PDF in the ISO

Refs: #1695
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @slaperche-scality

* @gdemonet

* @carlito-scality

* @sayf-eddine-scality

* @nootal

* @thomasdanan

* @alexandre-allard-scality

* @lucieleonard

* @zempashi

* @monpote

* @jbertran

* @cmonfort

* @nicolast

* @teddyandrieux

gdemonet commented on buildchain/buildchain/docs.py:
```diff
@@ -0,0 +1,98 @@
+            ),
+            'actions': [['tox', '-e', 'docs', '--', target.command]],
+            'targets': [target.target],
+            'file_dep': list(utils.git_ls('docs')),
```
Neat trick, I'll use something like this for the Makefile in `examples/example-solution`

gdemonet commented on buildchain/buildchain/docs.py:
```diff
@@ -0,0 +1,98 @@
+        """Delete the build sub-directory for the given target."""
+        return lambda: coreutils.rm_rf(target.target.parent)
+
+    doc_targets = (
```
Fine for now, but could be useful to be able to choose, no? If developing docs, I guess `tox -e docs -- livehtml` is more comfortable, but `./doit.sh doc_html` could be cool as well IMHO

slaperche-scality replied:
I thought about supporting `livehtml`, but with `doit.sh auto doc:html` you'll have auto-recompilation as well no?

NicolasT replied:
That's not the only thing `livehtml` does: it recompiles, but also embeds some JS in the result such that when recompiling, the server can send an event to the browser to trigger a reload.

slaperche-scality replied:
Yeah, and it spawn a small webserver to serve the doc as well I think.

But meh, close enough :stuck_out_tongue:  (there is still `tox` if you really want to), or I could add a target `doc:live` (or `doc:auto`, or `doc:watch`), won't be hard.

I'll see once I have something working on the CI

slaperche-scality replied:
Done

slaperche-scality committed:
build/doc: allow to reuse pre-build doc

In the CI, the documentation is built by another container/step and
during the build step we want to reuse this artifact instead of
re-building it ourself.

This commit introduce a new config knob to check if we are running in
the CI or not.

This is is a kinda internal settings, we don't document it outside of
the code.

Refs: #1695
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
ci: reuse `docs` artifacts for the `build` step

We now embed the PDF documentation in the ISO.
Since we are already building the documentation in a dedicated step,
let's reuse the output in the `build` step.

This reuse allows to keep distinct containers for doc and ISO building.

Closes: #1695
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/doc: add a task wrapping `livehtml`

Useful task when hacking on the documentation.

Refs: #1695
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on eve/main.yml:
```diff
@@ -123,6 +135,8 @@ stages:
               source metalk8s-%(prop:product_version_prev)s/VERSION &&
               echo %(prop:product_version_prev)s.$VERSION_PATCH$VERSION_SUFFIX'
+
+      - ShellCommand: *download_pdf_doc
```
So now we'd end up with a 'previous version' build/ISO containing the docs of the 'current version'?

slaperche-scality replied:
Yeah, I guess for upgrade/downgrade we don't care about the embedded doc anyway.
It's just a matter of having a build that works.

Now if we have a simple way to retrieve the matching doc, why not.
If it's more involved then I'm not sure it's worth the effort.

NicolasT replied:
Can't we simply not embed any doc in the resulting ISO, then?

slaperche-scality replied:
But we always want the doc in the ISO, no?
Why should me make it optional? Only to workaround a quirk in our CI workflow?

This ISO is only built for testing upgrade/downgrade, it won't be uploaded anywhere to be reused.
I would say that embdedding a wrong PDF in a transient ISO is a lesser evil than making doc embedding optional in the buildchain (which could bite us later it the codepath is triggered by error when we build the real ISO we ship)

NicolasT reviewed (approved):
Didn't test, but LGTM.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality

slaperche-scality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
